servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpDeserializer2.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright © 2021 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.http.api;
+
+import io.servicetalk.buffer.api.Buffer;
+import io.servicetalk.buffer.api.BufferAllocator;
+
+/**
+ * A factory to address deserialization concerns for HTTP request/response payload bodies.
+ * @param <T> The type of objects to deserialize.
+ */
+public interface HttpDeserializer2<T> {
+    /**
+     * Deserialize a single {@link Object} into a {@link T}.
+     * @param headers The {@link HttpHeaders} associated with the {@code payload}.
+     * @param allocator Used to allocate intermediate {@link Buffer}s if required.
+     * @param payload The {@link Object} to deserialize. The contents are assumed to be in memory, otherwise this method
+     * may block.
+     * @return The result of the deserialization.
+     */
+    T deserialize(HttpHeaders headers, BufferAllocator allocator, Buffer payload);
```

Comment:
I started with the approach of using `HttpDeserializer` and `HttpSerializer`, as these are the desired names longer term, but the usage of this API on `HttpRequest` and `HttpResponse` is such that it should only be the aggregated methods. The new implementations of these APIs don't make sense for the streaming paths (e.g. no framing is specified) so I could implement these to fail with UnsupportedOperationException, but then users may accidentally pass implementations to streaming APIs and get exceptions at runtime. Having a distinct API avoids passing impls to the wrong method and partially implemented interface, but negative is that the method name isn't desirable long term. However since it is intended to be only "temporary" (likely 3+ releases bcz these changes are central) maybe this is OK... wdyt?

Reply:
Since the interface is intended to be used only temporarily the naming is not a problem.